graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/util/DirectedGraph.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.graphql.schema.schemafirst.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import net.jcip.annotations.NotThreadSafe;
+
+/** A basic directed graph implementation to perform topological sorts. */
+@NotThreadSafe
+public class DirectedGraph<VertexT> {
+
+  // We need to keep track of the predecessor count. For simplicity, use a map to store it
+  // alongside the vertices.
+  private final Map<VertexT, Integer> vertices;
+  private final Multimap<VertexT, VertexT> adjacencyList;
+  private boolean wasSorted;
+
+  public DirectedGraph(Collection<VertexT> vertices) {
+    this.vertices = Maps.newLinkedHashMapWithExpectedSize(vertices.size());
+    this.adjacencyList = LinkedHashMultimap.create();
+
+    for (VertexT vertex : vertices) {
+      this.vertices.put(vertex, 0);
+    }
+  }
+
+  @VisibleForTesting
+  @SafeVarargs
+  DirectedGraph(VertexT... vertices) {
+    this(Arrays.asList(vertices));
+  }
+
+  /**
+   * this assumes that {@code from} and {@code to} were part of the vertices passed to the
+   * constructor
+   */
+  public void addEdge(VertexT from, VertexT to) {
+    Preconditions.checkArgument(vertices.containsKey(from) && vertices.containsKey(to));
+    adjacencyList.put(from, to);
+    vertices.put(to, vertices.get(to) + 1);
```

Comment:
For the layman can you please add a comment about what the increment means here?